model/quic-socket-tx-pfifo-scheduler.cc

```diff
@@ -0,0 +1,301 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * Authors: Alvise De Biasio <alvise.debiasio@gmail.com>
+ *          Federico Chiariotti <chiariotti.federico@gmail.com>
+ *          Michele Polese <michele.polese@gmail.com>
+ *          Davide Marcato <davidemarcato@outlook.com>
+ *          
+ */
+
+#include "quic-socket-tx-pfifo-scheduler.h"
+
+#include <algorithm>
+#include <iostream>
+#include <sstream>
+#include "ns3/simulator.h"
+
+#include "ns3/packet.h"
+#include "ns3/log.h"
+#include "ns3/abort.h"
+#include "quic-subheader.h"
+#include "quic-socket-base.h"
+
+namespace ns3 {
+
+NS_LOG_COMPONENT_DEFINE("QuicSocketTxPFifoScheduler");
+
+NS_OBJECT_ENSURE_REGISTERED(QuicSocketTxPFifoScheduler);
+
+TypeId QuicSocketTxPFifoScheduler::GetTypeId(void) {
+	static TypeId tid = TypeId("ns3::QuicSocketTxPFifoScheduler").SetParent<
+			QuicSocketTxScheduler>().SetGroupName("Internet").AddConstructor<
+			QuicSocketTxPFifoScheduler>().AddAttribute("RetxFirst",
+			"Prioritize retransmissions regardless of stream",
+			BooleanValue(false),
+			MakeBooleanAccessor(&QuicSocketTxPFifoScheduler::m_retxFirst),
+			MakeBooleanChecker());
+	return tid;
+}
+
+QuicSocketTxPFifoScheduler::QuicSocketTxPFifoScheduler() :
+		QuicSocketTxScheduler(), m_appSize(0), m_retxFirst(false) {
+	m_appList = QuicTxPacketList();
+}
+
+QuicSocketTxPFifoScheduler::QuicSocketTxPFifoScheduler(
+		const QuicSocketTxPFifoScheduler &other) :
+		QuicSocketTxScheduler(other), m_appSize(other.m_appSize), m_retxFirst(
+				other.m_retxFirst) {
+	m_appList = other.m_appList;
+}
+
+QuicSocketTxPFifoScheduler::~QuicSocketTxPFifoScheduler(void) {
+	m_appList = QuicTxPacketList();
+	m_appSize = 0;
+}
+
+void QuicSocketTxPFifoScheduler::Print(std::ostream &os) const {
+	NS_LOG_FUNCTION(this);
+	QuicTxPacketList temp = QuicTxPacketList(m_appList);
+	std::stringstream as;
+
+	while (temp.size() > 0) {
+		PriorityTxItem pItem = temp.top();
+		pItem.item->Print(as);
+		temp.pop();
+	}
+
+	os << Simulator::Now().GetSeconds() << "\nApp list: \n" << as.str()
+			<< "\n\nCurrent Status: \nApplication Size = " << m_appSize;
+}
+
+void QuicSocketTxPFifoScheduler::Add(Ptr<QuicSocketTxItem> item, bool retx) {
+	NS_LOG_FUNCTION(this << item);
+
+	if (retx) {
+		if (m_retxFirst) {
+			NS_LOG_INFO("Adding retransmitted packet with highest priority");
+			m_appList.push(PriorityTxItem(0, 0, item));
+			m_appSize += item->m_packet->GetSize();
+		} else {
+			uint32_t dataSizeByte = item->m_packet->GetSize();
+			QuicSubheader sub;
+			item->m_packet->PeekHeader(sub);
+			if (sub.GetSerializedSize() + sub.GetLength()
+					< item->m_packet->GetSize()) {
+
+				NS_LOG_INFO(
```

Comment:
Does this disgregation has features that depend on the implementation of PFIFO or EDF? If not, I would encapsulate this code in a function for the parent class, to avoid code duplication

Reply:
Maybe the whole Add method could be for the parent and not overridden, maybe you can define a GetPriority method that behaves differently according to PFIFO EDF etc